data/tutorials/language/1ms_01_functors.md

```diff
@@ -8,183 +8,392 @@ category: "Module System"
+  |> IterPrint.f
+```
+
+The dependency `List` is _injected_ when compiling the module `Funkt`. Observe that the code using `IterPrint` is unchanged. Check the program's behaviour using `dune exec funkt < dune`.
+
+**Replacing a Dependency**
+
+Now, replacing the implementation of `iter` inside `IterListPrint` is no longer a refactoring; it is another functor application with another dependency. Here, `Array` replaces `List`:
+
+**`funkt.ml`**
+```ocaml
+module StringSet = Set.Make(String)
+module IterPrint = IterPrint.Make(Array)
+
+let _ =
+  stdin
+  |> In_channel.input_lines
+  |> List.concat_map Str.(split (regexp "[ \t.,;:()]+"))
+  |> StringSet.of_list
+  |> StringSet.elements
+  |> Array.of_list
+  |> IterPrint.f
+```
+
+Check the program's behaviour using `dune exec funkt < dune`.
+
+**Note**: The functor `IterPrint.Make` returns a module that exposes the type of the injected dependency (here first `List.t` then `Array.t`). That's why a `with type` constraint is needed. If the dependency was an _implementation detail_ that is not exposed in the signature of the initial version of `IterMake` (i.e., in the type of `IterMake.f`), that constraint wouldn't be needed, and the call site of `IterPrint.f` would be unchanged when injecting another dependency.
+
+## Write a Functor to Extend Modules
+
+In this section, we define a functor to extend several modules in the same way. This is the same idea as in the [Extending a Module with a Standard Library Functor](#extending-a-module-with-a-standard-library-functor), except we write the functor ourselves.
+
+Create a fresh directory with the following files:
+
+**`dune-project`**
+```lisp
+(lang dune 3.7)
+```
+**`dune`**
+```lisp
+(library (name scanLeft))
+```
+
+**`scanLeft.ml`**
+```ocaml
+module type LeftFoldable = sig
+  type 'a t
+  val fold_left : ('b -> 'a -> 'b) -> 'b -> 'a t -> 'b
+  val of_list : 'a list -> 'a t
+end
+
+module type S = sig
+  type 'a t
+  val scan_left : ('b -> 'a -> 'b) -> 'b -> 'a t -> 'b t
+end
+
+module Make(F: LeftFoldable) : S with type 'a t := 'a F.t = struct
+  let scan_left f b u =
+    let f (b, u) a = let b' = f b a in (b', b' :: u) in
+    u |> F.fold_left f (b, []) |> snd |> List.rev |> F.of_list
+end
+```
+
+Run the `dune utop` command. Once inside the toplevel, enter the following commands. For brievety, the output of the first two toplevel commands is not shown here.
+```ocaml
+# module Array = struct
+    include Stdlib.Array
+    include ScanLeft.Make(Stdlib.Array)
+  end;;
+
+# module List = struct
+    include List
+    include ScanLeft.Make(struct
+      include List
+      let of_list = Fun.id
+    end)
+  end;;
+
+# Array.init 10 Fun.id |> Array.scan_left ( + ) 0;;
+- : int array = [|0; 1; 3; 6; 10; 15; 21; 28; 36; 45|]
+
+# List.init 10 Fun.id |> List.scan_left ( + ) 0;;
+- : int list = [0; 1; 3; 6; 10; 15; 21; 28; 36; 45]
+```
+
+Modules `Array` and `List` appear augmented with `Array.scan_left` and `List.scan_left`.
+
+## Conclusion
```

Comment:
Should we complete this section or remove it?

Reply:
It needs to be completed, we need some conclusion